Zotero/Controllers/OpenItemsController.swift

```diff
@@ -0,0 +1,481 @@
+                    let currentItemMenu = UIMenu(title: L10n.Accessibility.Pdf.currentItem, options: [.displayInline], children: currentItemActions)
+                    let currentItemElement = UIMenu(title: rItem.displayTitle, image: item.kind.icon, children: [currentItemMenu])
+                    elements.append(currentItemElement)
+                } else {
+                    let itemAction = UIAction(title: rItem.displayTitle, image: item.kind.icon) { [weak self] _ in
+                        guard let self, let presenter = openItemPresenterProvider() else { return }
+                        restore(item, using: presenter)
+                        completion(true, false)
+                    }
+                    elements.append(itemAction)
+                }
+            }
+
+            let closeAllAction = UIAction(title: L10n.Accessibility.Pdf.closeAllOpenItems, image: .init(systemName: "xmark.square")) { [weak self] _ in
+                guard let self else { return }
+                setItems([], for: sessionIdentifier, validate: false)
+                openItemPresenterProvider()?.showItem(with: nil)
+                completion(true, true)
+            }
+            let closeAllElement = UIMenu(options: [.displayInline], children: [closeAllAction])
+            elements.append(closeAllElement)
+
+            elementProvider(elements)
+        }
+    }
+    
+    // MARK: Helper Methods
+    private func itemsSortedByUserOrder(for sessionIdentifier: String) -> [Item] {
+        getItems(for: sessionIdentifier).sorted(by: { $0.userIndex < $1.userIndex })
+    }
+
+    private func itemsSortedByLastOpen(for sessionIdentifier: String) -> [Item] {
+        getItems(for: sessionIdentifier).sorted(by: { $0.lastOpened > $1.lastOpened })
+    }
+
+    private func filterValidItemsWithRItem(_ items: [Item]) -> [(Item, RItem)] {
+        var itemTuples: [(Item, RItem)] = []
+        do {
+            try dbStorage.perform(on: .main) { coordinator in
+                for item in items {
+                    switch item.kind {
+                    case .pdf(let libraryId, let key), .note(let libraryId, let key):
+                        do {
+                            let rItem = try coordinator.perform(request: ReadItemDbRequest(libraryId: libraryId, key: key))
+                            itemTuples.append((item, rItem))
+                        } catch let itemError {
+                            DDLogError("OpenItemsController: can't load item \(item) - \(itemError)")
+                        }
+                    }
+                }
+            }
+        } catch let error {
+            DDLogError("OpenItemsController: can't load multiple items - \(error)")
+        }
+        return itemTuples
+    }
+
+    private func filterValidItems(_ items: [Item]) -> [Item] {
+        filterValidItemsWithRItem(items).map { $0.0 }
+    }
+
+    private func loadPresentation(for item: Item) -> Presentation? {
+        switch item.kind {
```

Comment:
Here you could do the `do-catch` block with `try dbStorage.perform(on: .main) { coordinator in` in this function and pass `coordinator` to `loadPDFPresentation` and `loadNotePresentation?` and both would be throwable, something like: 
```
func loadPDFPresentation(key: String, libraryId: LibraryIdentifier, coordinator: DbCoordinator) throws -> Presentation?
func loadNotePresentation(key: String, libraryId: LibraryIdentifier, coordinator: DbCoordinator) throws -> Presentation?
```

and then the child functions would be a bit cleaner without repeating `do-catch` blocks.

Reply:
Check https://github.com/zotero/zotero-ios/pull/829/commits/ae2376422f3f11d32545767dac3d7b35913fac7a

Reply:
I don't really mind either approach, but why did you move back to previous format? Did it cause any issues?

Reply:
I rewrote the part to add a completion, and I guess I followed the same steps 😅 Check https://github.com/zotero/zotero-ios/pull/829/commits/a607962689c73afac08a839640c22cc3a5d4d0d0